packages/expo-audio/android/src/main/java/expo/modules/audio/AudioModule.kt

```diff
@@ -1,10 +1,500 @@
+import expo.modules.core.interfaces.services.UIManager
+import expo.modules.interfaces.permissions.Permissions
+import expo.modules.kotlin.Promise
+import expo.modules.kotlin.exception.Exceptions
+import expo.modules.kotlin.functions.Queues
 import expo.modules.kotlin.modules.Module
 import expo.modules.kotlin.modules.ModuleDefinition
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import okhttp3.OkHttpClient
+import kotlin.math.min
+
+@androidx.annotation.OptIn(androidx.media3.common.util.UnstableApi::class)
+class AudioModule : Module(), LifecycleEventListener {
+  private val activity: Activity
+    get() = appContext.activityProvider?.currentActivity ?: throw Exceptions.MissingActivity()
+  private lateinit var audioManager: AudioManager
+  private val context: Context
+    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()
+
+  private val players = mutableMapOf<String, AudioPlayer>()
+  private var appIsPaused = false
+  private var staysActiveInBackground = false
+  private var audioEnabled = true
+  private var audioInterruptionMode = AudioInterruptionMode.DUCK_OTHERS
+  private var shouldRouteThroughEarpiece = false
+
+  private enum class AudioInterruptionMode {
+    DO_NOT_MIX,
+    DUCK_OTHERS
+  }
 
-class AudioModule : Module() {
   override fun definition() = ModuleDefinition {
     Name("ExpoAudio")
+
+    OnCreate {
+      audioManager = appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as AudioManager
+      val uiManager = appContext.legacyModule<UIManager>()
+      uiManager?.registerLifecycleEventListener(this@AudioModule)
+    }
+
+    AsyncFunction("setAudioModeAsync") { mode: AudioMode ->
+      staysActiveInBackground = mode.shouldPlayInBackground
+      shouldRouteThroughEarpiece = mode.shouldRouteThroughEarpiece ?: false
+      if (shouldRouteThroughEarpiece) {
+        updatePlaySoundThroughEarpiece(true)
+      }
+    }
+
+    AsyncFunction("setIsAudioActiveAsync") { enabled: Boolean ->
+      audioEnabled = enabled
+      if (!enabled) {
+        appContext.mainQueue.launch {
+          players.values.forEach {
+            if (it.player.isPlaying) {
+              it.player.pause()
+            }
+          }
+        }
+      }
+    }
+
+    AsyncFunction("requestRecordingPermissionsAsync") { promise: Promise ->
+      Permissions.askForPermissionsWithPermissionsManager(appContext.permissions, promise, Manifest.permission.RECORD_AUDIO)
+    }
+
+    AsyncFunction("getRecordingPermissionsAsync") { promise: Promise ->
+      Permissions.getPermissionsWithPermissionsManager(appContext.permissions, promise, Manifest.permission.RECORD_AUDIO)
+    }
+
+    OnDestroy {
+      val uiManager = appContext.legacyModule<UIManager>()
+      uiManager?.unregisterLifecycleEventListener(this@AudioModule)
+    }
+
+    Class(AudioPlayer::class) {
+      Constructor { source: AudioSource? ->
+        val builder = OkHttpClient().newBuilder()
```

Comment:
Clients should be shared (check https://square.github.io/okhttp/5.x/okhttp/okhttp3/-ok-http-client/index.html). Creating a new one for each session isn't a good idea. There should be a construction that allows you to customize things like headers.